### PR TITLE
Add option to print word-level scores

### PR DESCRIPTION
**Pull request by snukky**
This adds two options to marian-decoder:
- `--word-scores` displays (not normalized) word-level scores calculated from the path score.
- `--no-spm-decode` keeps output BPE-segmented, available only if Marian is compiled with SentencePiece. 

The latter may be handy if used with `--word-scores`/`--alignment` to get 1:1 mapping between output tokens and scores/alignment.

Examples:
```
echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores
Tohle je test. ||| WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176

echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores --n-best
0 ||| Tohle je test. ||| F0= -1.77547 F1= -1.91984 WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176 ||| -0.739061
0 ||| Tohle je zkouška. ||| F0= -2.09314 F1= -2.03168 WordScores= -1.51923 -0.21951 -1.92797 -0.23666 -0.22146 ||| -0.824964
0 ||| To je test. ||| F0= -3.32969 F1= -3.23537 WordScores= -4.19787 -0.27764 -1.61160 -0.25529 -0.22266 ||| -1.31301

echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores --n-best --no-spm-decode
0 ||| ▁Tohle ▁je ▁test . </s> ||| F0= -1.77547 F1= -1.91984 WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176 ||| -0.739061
0 ||| ▁Tohle ▁je ▁zkouška . </s> ||| F0= -2.09314 F1= -2.03168 WordScores= -1.51923 -0.21951 -1.92797 -0.23666 -0.22146 ||| -0.824964
0 ||| ▁To ▁je ▁test . </s> ||| F0= -3.32969 F1= -3.23537 WordScores= -4.19787 -0.27764 -1.61160 -0.25529 -0.22266 ||| -1.31301

echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores --no-spm-decode --alignment
▁Tohle ▁je ▁test . </s> ||| 1-0 5-1 5-2 5-3 5-4 ||| WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176
```

### CODE REVIEW

File: src/data/sentencepiece_vocab.cpp [hunk 1]
```diff
@@ -36,6 +36,8 @@ class SentencePieceVocab : public IVocab {
   std::mt19937 generator_;
   std::uniform_int_distribution<int> randInt_; // from 0 to INT_MAX
 
+  bool keepEncoded_{false};
```
Comment (emjotde) [inline src/data/sentencepiece_vocab.cpp]: Please add a comment what this does.

File: src/data/sentencepiece_vocab.cpp [hunk 2]
```diff
@@ -223,11 +227,18 @@ class SentencePieceVocab : public IVocab {
 
   std::string decode(const Words& sentence, bool /*ignoreEOS*/) const override {
     std::string line;
-    // convert vector of Word to vector of int
-    std::vector<int> spmSentence; spmSentence.reserve(sentence.size());
-    for (auto&& word : sentence)
-      spmSentence.push_back(word.toWordIndex());
-    spm_->Decode(spmSentence, &line);
+    if(keepEncoded_) {
```
Comment (emjotde) [inline src/data/sentencepiece_vocab.cpp]: Comment please.

File: src/translator/output_printer.cpp [hunk 3]
```diff
@@ -26,4 +28,12 @@ std::string OutputPrinter::getAlignment(const Ptr<Hypothesis>& hyp) {
   }
 }
 
+std::string OutputPrinter::getWordScores(const Ptr<Hypothesis>& hyp) {
```
Comment (emjotde) [inline src/translator/output_printer.cpp]: Comment what these are.

**Comment by emjotde** [comment]
Generally approved. I am a bit unsure about the missing `|||` between feature functions and word scores. Word scores cannot be used for MERT or similar optimizations.

File: src/data/sentencepiece_vocab.cpp (hunk 2 above)
Comment (snukky) [reply src/data/sentencepiece_vocab.cpp]: Done

File: src/data/sentencepiece_vocab.cpp (hunk 1 above)
Comment (snukky) [reply src/data/sentencepiece_vocab.cpp]: Added

File: src/translator/output_printer.cpp (hunk 3 above)
Comment (snukky) [reply src/translator/output_printer.cpp]: I added a more detailed comment to the function generating the scores

**Comment by snukky** [comment]
Added comments and changed the placement of word-level scores in the n-best list to a separate field before feature function scores.
With word alignment, the order is as follows: id, sentence, word alignment, word scores, feature functions, total score

**Pull request merged**